src/OpenTelemetry.Exporter.Prometheus.Shared/PrometheusExporter.cs

```diff
@@ -28,32 +28,16 @@ public class PrometheusExporter : BaseExporter<Metric>, IPullMetricExporter
     {
         internal const string HttpListenerStartFailureExceptionMessage = "PrometheusExporter http listener could not be started.";
         internal readonly PrometheusExporterOptions Options;
-        private readonly PrometheusExporterHttpServer metricsHttpServer;
         private Func<int, bool> funcCollect;
         private Func<Batch<Metric>, ExportResult> funcExport;
-        private bool disposed;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="PrometheusExporter"/> class.
         /// </summary>
         /// <param name="options">Options for the exporter.</param>
-        public PrometheusExporter(PrometheusExporterOptions options)
+        internal PrometheusExporter(PrometheusExporterOptions options)
```

Comment:
One extra thing I just thought of...

Anything in "Shared" project which is public and linked into some other project could become a problem. For example if users reference both projects (HttpListener & AspNetCore) and then they do `new PrometheusExporter(...)` the compiler won't know which one to use. Might not be something to worry about. We _could_ get around that by switching the namespaces based on compiler constants. Something like this...

```csharp
#if PROMETHEUS_ASPNETCORE
namespace OpenTelemetry.Exporter.Prometheus.AspNetCore
#elif PROMETHEUS_ASPNETCORE_HTTPLISTENER
namespace OpenTelemetry.Exporter.Prometheus.HttpListener
#else
namespace OpenTelemetry.Exporter.Prometheus.Shared
#endif
```

Then define `PROMETHEUS_ASPNETCORE` in the aspnetcore project csproj, `PROMETHEUS_ASPNETCORE_HTTPLISTENER` in the httplistener project csproj.

```xml
  <PropertyGroup>
    <DefineConstants>$(DefineConstants);PROMETHEUS_ASPNETCORE</DefineConstants>
  </PropertyGroup>
```

Reply:
This is a very good point; I actually encountered something similar at first when I was trying to put tests for both HttpListener and AspNetCore in the same project. But I later figured out putting them in the same project may not be a common use case. :-)

Reply:
Updated the ctor to be public.

Reply:
@alanwest @cijothomas @utpilla @reyang 

Anyone have an opinion on what we should do here? To summarize the situation...

* OpenTelemetry.Exporter.Prometheus.Shared.PrometheusExporter is linked into OpenTelemetry.Exporter.Prometheus.AspNetCore & OpenTelemetry.Exporter.Prometheus.HttpListener as `public OpenTelemetry.Exporter.PrometheusExporter`.
* It is unlikely(?) users will ref both packages, but shipping like this could create some problems down the road because 2 assemblies are exposing the same public type in the same namespace.

3 potential solutions I can think of:

* Make _everything_ linked `internal`. Today `PrometheusExporter` is `public` so this could upset some users.
* Actually ship `OpenTelemetry.Exporter.Prometheus.Shared` as a package and just reference it normally. This means users have another assembly to deal with, may or may not be a concern.
* Use the namespace trick from above so that each linked copy ends up distinct in its final assembly.

Reply:
Is there any need of exposing `PrometheusExporterOptions` at all? If only `ScrapeEndpointPath` is required, then we can make it separately in each of the Exporters.

Reply:
I have used the difference ns trick in ApplicationInsights https://github.com/microsoft/ApplicationInsights-dotnet/blob/main/NETCORE/src/Shared/Extensions/ApplicationInsightsServiceOptions.cs#L1-L4.

There was some issue about this.. trying to recollect what that is. will post here, if i find the actual issue..

Reply:
> Is there any need of exposing `PrometheusExporterOptions` at all? If only `ScrapeEndpointPath` is required, then we can make it separately in each of the Exporters.

I kinda like the idea of exposing PrometheusExporterOptions mainly because it provides a clear semantic distinction between listener and prometheusExporter. 
And since we anyway put the shared code in one project, referencing it should not be a big problem. From the consumer's perspective, I don't think keeping these 2 separate options would be too complicated, rather it provides semantic clarity.

Reply:
Sharing my position - I think all shared files should only have `internal` / `private` classes, if the intention is to share the types, it should be a binary/package dependency (or text template files, which have their own set of problems).

Reply:
> 3 potential solutions I can think of:
> 
> * Make _everything_ linked `internal`. Today `PrometheusExporter` is `public` so this could upset some users.
> * Actually ship `OpenTelemetry.Exporter.Prometheus.Shared` as a package and just reference it normally. This means users have another assembly to deal with, may or may not be a concern.
> * Use the namespace trick from above so that each linked copy ends up distinct in its final assembly.

@CodeBlanch I agree with your ordering. The 1st one sounds the best to me. The 2nd one doesn't sound as good, because `OpenTelemetry.Exporter.Prometheus.Shared` is like an overkill (it's not a super complex project, it makes the package dependency more complicated, it's just hard to justify why we need to have such one more package).

Reply:
Based on the feedback - I've updated the `PrometheusExporter` to be internal but will keep this conversation open for future reference.

Reply:
I would also agree for shared packages to only contain `internal` things but did we decide to go that route?

We still have `PrometheusExporterOptions` class marked as `public` in the shared project.

There is also the `public` extension method `AddPrometheusExporter(Action<PrometheusExporterOptions>)` in the shared project. If this extension method is only used the AspNetCore project, wouldn't it better to move it there?

Reply:
> We still have `PrometheusExporterOptions` class marked as `public` in the shared project.

Right, I've updated to use `$<DefinedConstants>` to get away with it.
 
> There is also the `public` extension method `AddPrometheusExporter(Action<PrometheusExporterOptions>)` in the shared project. If this extension method is only used the AspNetCore project, wouldn't it better to move it there?

Yes, `.AddPrometheusExporter()` is move away from the shared project and now resides in AspNetCore project.

Reply:
Hey @alanwest @cijothomas @CodeBlanch @Kielek @reyang @utpilla:
I've tried out another projects layout given the suggestion:
https://github.com/open-telemetry/opentelemetry-dotnet/pull/3497#discussion_r931572797
with this commit: https://github.com/open-telemetry/opentelemetry-dotnet/pull/3497#discussion_r931572797

With this way, it cleans up the namespaces a little bit and works around the public api documentation issue as was mentioned here: https://github.com/open-telemetry/opentelemetry-dotnet/pull/3430#discussion_r931522363

Please let me know how you feel about this change! Thank you.

Reply:
> Hey @alanwest @cijothomas @CodeBlanch @Kielek @reyang @utpilla: I've tried out another projects layout given the suggestion: [#3497 (comment)](https://github.com/open-telemetry/opentelemetry-dotnet/pull/3497#discussion_r931572797) with this commit: [#3497 (comment)](https://github.com/open-telemetry/opentelemetry-dotnet/pull/3497#discussion_r931572797)

With **this** commit https://github.com/open-telemetry/opentelemetry-dotnet/pull/3430/commits/3b562b97d3902e59acfc8886e69d977ae83e8c1c?

Reply:
> > Hey @alanwest @cijothomas @CodeBlanch @Kielek @reyang @utpilla: I've tried out another projects layout given the suggestion: [#3497 (comment)](https://github.com/open-telemetry/opentelemetry-dotnet/pull/3497#discussion_r931572797) with this commit: [#3497 (comment)](https://github.com/open-telemetry/opentelemetry-dotnet/pull/3497#discussion_r931572797)
> 
> With **this** commit [3b562b9](https://github.com/open-telemetry/opentelemetry-dotnet/commit/3b562b97d3902e59acfc8886e69d977ae83e8c1c)?

yes, my bad.

Reply:
Please disregard my previous comments about the 3b562b9 commit.

I've made a new PR to try out 2 project setup for prometheus here:
https://github.com/open-telemetry/opentelemetry-dotnet/pull/3503

Please take a look.